### PR TITLE
Added ELF memory snapshot function

### PR DESCRIPTION
**Pull request by eric-c-wood**
ELF memory snapshot function was added to allow the use of various memory forensic toolkits.

### CODE REVIEW

File: src/go/api/vm/util.go [hunk 1]
```diff
@@ -172,3 +172,10 @@ func (this cancelableReader) Read(p []byte) (int, error) {
 		return this.r.Read(p)
 	}
 }
+
+func getTimestamp() string {
+
+	refTime := time.Now()
+	return fmt.Sprintf("%d%02d%02d_%02d%02d", refTime.Year(), refTime.Month(),
+		refTime.Day(), refTime.Hour(), refTime.Minute())
```
Comment (activeshadow) [inline src/go/api/vm/util.go:180]: @eric-c-wood for consistency's sake, I'd prefer to see the `time.Format` function used here. To reproduce the same timestamp you're producing here, you could use `refTime.Format("20060102_1500")`.

**Review by activeshadow — CHANGES_REQUESTED** [review]
@eric-c-wood the functionality this PR provides is awesome, and is something other users have expressed as needing also. Thanks for putting it together! Just a few comments/questions before we can get it merged in.

File: src/go/api/vm/util.go (hunk 1 above)
Comment (eric-c-wood) [reply src/go/api/vm/util.go:180]: @activeshadow will fix.

File: src/go/cmd/vm.go [hunk 2]
```diff
@@ -397,6 +397,48 @@ func newVMCaptureCmd() *cobra.Command {
...
 }
 
+func newVMMemorySnapshotCmd() *cobra.Command {
+	desc := `Create an ELF memory snapshot of the VM
+	
+  Used to create an ELF memory snapshot for a running virtual machine
+  that is compatible with memory forensic toolkits Volatility and Google's Rekall.`
+
+	cmd := &cobra.Command{
+		Use:   "memorysnapshot <experiment name> <vm name> <snapshot file path>",
```
[the remark this answers is not in the document]
Comment (eric-c-wood) [reply src/go/cmd/vm.go:407]: @activeshadow will fix

File: src/go/api/vm/vm.go [hunk 3]
```diff
@@ -1058,3 +1058,118 @@ func CommitToDisk(expName, vmName, out string, cb func(float64)) (string, error)
...
+	cmd.Columns = nil
+	cmd.Filters = nil
+	cmd.Command = fmt.Sprintf("%s shell mkdir -p %s", cmdPrefix, filepath.Dir(out))
+
+	if err := mmcli.ErrorResponse(mmcli.Run(cmd)); err != nil {
+		return "", fmt.Errorf("ensuring experiment files directory exists: %w", err)
+	}
+	// ***** BEGIN: MEMORY SNAPSHOT VM *****
+
+	qmp := fmt.Sprintf(`{ "execute": "dump-guest-memory", "arguments": { "protocol": "file:%s", "paging": false, "format": "elf" , "detach": true} }`, out)
```
[the remark this answers is not in the document]
Comment (eric-c-wood) [reply src/go/api/vm/vm.go:1116]: @activeshadow From reading the QMP `dump-guest-memory` command documentation, there is no requirement to pause/stop the VM prior to executing the `dump-guest-memory` command.  Consequently, the `dump-guest-memory` command can be executed on a running VM.

**Comment by eric-c-wood** [comment]
@activeshadow  I believe that I made all of your recommended code changes.  Please let me know if I missed anything.  Thanks for the feedback.  I will make a note of all the issues to avoid in any future PR requests.

File: src/go/web/handlers.go [hunk 4]
```diff
@@ -2117,6 +2110,153 @@ func CommitVM(w http.ResponseWriter, r *http.Request) {
...
+	var (
+		ctx      = r.Context()
+		role     = ctx.Value("role").(rbac.Role)
+		vars     = mux.Vars(r)
+		exp      = vars["exp"]
+		name     = vars["name"]
+		fullName = exp + "_" + name
+	)
+
+	if !role.Allowed("vms/memorySnapshot", "create", fullName) {
```
[the remark this answers is not in the document]
Comment (eric-c-wood) [reply src/go/web/handlers.go:2126]: @activeshadow I will take a look at the phenix-docs repo and add an appropriate PR.

File: src/go/web/handlers.go [hunk 5]
```diff
@@ -2117,6 +2110,131 @@ func CommitVM(w http.ResponseWriter, r *http.Request) {
...
+					broker.NewResource("experiment/vm/memorySnapshot", exp+"/"+name, "progress"),
+					marshalled,
+				)
+			}
+		}
+	}()
+
+	cb := func(s string) { status <- s }
+
+	if filename, err = vm.MemorySnapshot(exp, name, filename, cb); err != nil {
```
[the remark this answers is not in the document]
Comment (eric-c-wood) [reply src/go/web/handlers.go:2217]: @activeshadow The run time of the `vm.MemorySnapshot` is really a function of the disk write speed and the amount of memory being dumped.  For development, I was using an external USB hard drive and there was a noticeable delay (i.e. up to 45 secs) when dumping memory for multiple VMs. (i.e. < 3 GBs) When I used an internal hard drive, the delay for the same group of VMs was not even noticeable.  I have not yet tested with an SSD but I imagine that there would be a further decrease in delay.

With regards to the UI blocking, I believe the memory snapshot request similar to most web UI requests is asynchronous and does not really block the web UI.  While the UI can become sluggish, the user is still able to interact with the UI and is not locked out from further interactions.  On the other hand, we do block certain interactions with the VM that is having its memory dumped.  We also make use of a progress bar for each VM to give a status of the memory dumping progress.  

memory_progress2

**Review by activeshadow — APPROVED** [review]

File: src/go/web/handlers.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (activeshadow) [reply src/go/web/handlers.go:2217]: @eric-c-wood sounds good. I think this is ready to merge then!

**Comment by activeshadow** [comment]
@eric-c-wood can you resolve the merge conflicts that have come up from merging some other PRs today?

@glattercj do you want to work your QA magic on this PR, or are you good with it?

**Comment by eric-c-wood** [comment]
@activeshadow the two merge conflicts should now be resolved.

**Comment by glattercj** [comment]
@activeshadow LGTM

**Pull request merged**